File: 1_two_sum.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    static vector<int> twoSum(vector<int>& nums, int target) {
        vector<int> sol;
        int i=0,j=0;
        for(int x:nums){
            j=0;
            for(int y:nums){
                if(x+y==target&&i!=j){
                    sol.push_back(i);
                    sol.push_back(j);
                    break;
                }
                j++;
            }
            i++;
        }
        return sol;
    }
};


int main(){

    vector<int> nums;
    vector<int> sol;
    int a,target;
    while(cin>>a){
        nums.push_back(a);
        char c=getchar();
        if(c=='\n'){
            break;
        }
    }
    nums.pop_back();
    target=a;
    sol=Solution::twoSum(nums,target);
    cout<<'['<<sol[0]<<','<<sol[1]<<']';
    
}

File: 1_two_sum.py
class Solution:

    # O(nlogn) use hash table to store positions, sort the array and use two pointers
    def twoSum1(self, nums: list[int], target: int) -> list[int]:
        mp = {}
        equal = []
        for i,num in enumerate(nums):
            if num in mp and num + num == target:
                equal = [mp[num], i]
            else:
                mp[num] = i

        nums.sort()
        l, r = 0, len(nums) - 1

        while l < r - 1 and nums[l] + nums[r] != target:
            while l < r - 1 and nums[l] + nums[r] < target:
                l += 1
            while l < r - 1 and nums[l] + nums[r] > target:
                r -= 1
            
        return [mp[nums[l]], mp[nums[r]]] if nums[l] != nums[r] else equal

    # O(N)
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        mp = {}
        for i,num in enumerate(nums):
            if target - num in mp:
                return [mp[target-num], i]
            else:
                mp[num] = i
